chain/client/src/client.rs

```diff
@@ -898,12 +904,27 @@ impl Client {
     pub(crate) fn process_block_processing_artifact(
         &mut self,
         block_processing_artifacts: BlockProcessingArtifact,
+        apply_chunks_done_callback: DoneApplyChunkCallback,
     ) {
         let BlockProcessingArtifact { orphans_missing_chunks, blocks_missing_chunks, challenges } =
             block_processing_artifacts;
         // Send out challenges that accumulated via on_challenge.
         self.send_challenges(challenges);
-        // Request any missing chunks
+        // For any missing chunk, call the ShardsManager with it so that it may apply forwarded parts.
```

Comment:
It just occurred to me, maybe it flows better logically to call `self.shards_mgr.process_cached_chunk_forwards_for_header` in `preprocess_block` here https://github.com/near/nearcore/blob/master/chain/chain/src/chain.rs#L2219 and just call `self.check_blocks_with_missing_chunks` here after `preprocess_block` is called https://github.com/near/nearcore/blob/master/chain/chain/src/chain.rs#L1856.

I think it's the more natural way of logic. We try to process the chunk forwards message before we actually process the block. And processing these chunk forwards can unlock other blocks too, so we need to call the `check_blocks_with_missing_chunks` function later.

Reply:
Moving it to start_process_block_impl.